---
title: "lec18"
author: "Michael Nguyen"
date: "11/26/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Part 1
```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

```{r}
status()
```

Querying the GDC from R

We will typically start our interaction with the GDC by searching the resource to find data that we are interested in investigating further. In GDC speak this is called “Querying GDC metadata”. Metadata here refers to the extra descriptive information associated with the actual patient data (i.e. ‘cases’) in the GDC.


The are four main sets of metadata that we can query, namely projects(), cases(), files(), and annotations(). We will start with projects()


```{r}
projects <- getGDCprojects
head(projects)
```


```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

> (x$key == "TCGA-PAAD")





```{r}
x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))  # Makes lightblue nrow(X) of times
colvec[(x$key == "TCGA-PAAD")] <- "red"

# Plot with 'log' for y axis and **rotate labels with 'las'**
#par(___)  
# log scale to normalize the data
barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)
```
```{r}
?rep
```
```{r}
colvec
```


```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
```




```{r}
query <- GDCquery(project="TCGA-PAAD",
                  data.category="Transcriptome Profiling",
                  data.type="Gene Expression Quantification")

ans <- getResults(query)

head(ans)
```

```{r}
View(ans)
nrow(ans)
```




We could download these with standard R tools, or for larger data-sets such as this one, use the packages `transfer()` function, which uses the GDC transfer client (a separate command-line tool) to perform more robust data downloads.




## Variant analysis with R
Note we could go to the NCI-GDC web portal and enter the Advanced Search page and then construct a search query to find MAF format somatic mutation files for our ‘TCGA-PAAD’ project.

After some exploration of the website I came up with the following query: “cases.project.project_id in ["TCGA-PAAD"] and files.data_type in ["Masked Somatic Mutation"] and files.data_format in ["MAF"]”.

How many MAF files for the TCGA-PAAD project were found from this advanced web search?

Lets do the same search in R with the help of the TCGAbiolinks package function GDCquery_Maf(). For brevity we will focus on only one of the MAF files for this project, namely the MuTect2 workflow variant calls.



```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
```

```{r}
head(maf.file)
```


What argument could we use to write the MAF file into a csv document in your current working directory?

### MAF analysis
The MAF file contents is now stored as a dataframe and the maftools package workflow, which starts with a MAF file or dataframe, can proceed, starting with reading the pancreatic cancer MAF file.

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```


With the data now available as a maftools MAF object, a lot of functionality is available with little code. While the maftools package offers quite a few functions, here are a few highlights. Cancer genomics and bioinformatics researchers will recognize these plots:



## Plotting MAF summary.
We can use `plotmafSummary()` function to plot a summary of the maf object, which displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized by Variant_Classification. We can add either mean or median line to the stacked barplot to display average/median number of variants across the cohort.


```{r}
plotmafSummary(vars)
```



## Drawing oncoplots
A very useful summary representation of this data can be obtained via so-called oncoplots, also known as waterfall plots.

```{r}
oncoplot(maf = vars, top = 10)
```


### You might need to run the `oncoplot()` command in the R Console and then zoom the display to see the full plot (as it is rather large and may not appear initially in your Rmarkdown document before Knitting. Another option is to send your plot to a PNG or PDF plot device directly, for example:

```{r}
# Oncoplot for our top 10 most frequently mutated genes
pdf("oncoplot_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
dev.off()
```




# Part 2
WorkFlow
Fasta format


Background: To identify somatic mutations in a tumor, DNA from the tumor is sequenced and compared to DNA from normal tissue in the same individual using variant calling algorithms.

Comparison of tumor sequences to those from normal tissue (rather than ‘the human genome’) is important to ensure that the detected differences are not germline mutations.

To identify which of the somatic mutations leads to the production of aberrant proteins, the location of the mutation in the genome is inspected to identify non-synonymous mutations (i.e. those that fall into protein coding regions and change the encoded amino acid).



## Section 1. Protein sequences from healthy and tumor tissue

```{r}
library(bio3d)
```



```{r}
seqs <- read.fasta("lecture18_sequences.fa")
seqs
```

```{r}

```


```{r}
?bio3d
```

Your task is to identify tumor specific mutations that could potentially be used for vaccine development.

Q1: Identify sequence regions that contain all 9-mer peptides that are only found in the tumor. Hint: You will need to first identify the sites of mutation in the above sequences and **then extract the surrounding subsequence region.** This subsequence should encompass all possible 9-mers in the tumor derived sequence. In other words extract the subsequence from 8 residues before and 8 residues after all point mutations in the tumor sequence.

You could simply align the above sequences using your favorite alignment method and then visually identify differences and extract the required subsequences. However, this approach is error-prone and obviously will not scale to further examples (i.e. many additional proteins and 100s of mutations).

If you are motivated to use your new R coding skills to address this question (and you should be!) then you may be interested in the following example analysis. However, we encourage you to try on your own first before consulting this worked example.


```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method = "identity")
mutant.sites <- which(ide < 1)   # 1 being conserved!!
mutant.sites
```


```{r}
## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```


positions can change... so 9 variations of 9-mer
9-mers

## Making labelsor output seqs(one per mutant)
```{r}
mutant.names <- paste(seqs$ali["P53_wt", mutant.sites], 
                      mutant.sites, 
                     seqs$ali ["P53_mutant", mutant.sites])
mutant.names
```



**Now lets extract all 9-mer mutant encompassing sequences for each mutant site. This is equivalent to finding the sequence region eight residues before and eight residues after our mutation sites and outputting this subsequence to a new FASTA file.**


```{r}
# calculating postional identiy scores

ide <- conserv(seqs$ali, method = "identity")
pos <- which(ide <1)[1]
pos
subseq.pos <-(pos-8):(pos+8)
(pos-8):(pos+8)

```




```{r}
## Sequence positions surrounding each  mutant 
start.position <- mutant.sites -8
end.position <- mutant.sites +8

# Blank matrix to store ub-sequence
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs      #Sub-sequence
```


# 2 
HLA-A*02:01   SPLPSQAMLDLMLSPDD
HLA-A*68:01   DPGPDEAPWMPEAAPPV
HLA-B*07:02   YLDDRNTFVHSVVVPYE
HLA-B*35:01   ILTIITLEV--------

Finally lets output all these sequences to a FASTA file for further analysis with the IEDB HLA binding prediction website `http://tools.iedb.org/mhci/.`

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```


Q2: Identify 9-mer peptides in the identified sequence regions unique to the tumor that can be potentially presented to T cells. Hint: Use the IEDB HLA binding prediction server above to identify the top ranked 9-mer peptides for each patient HLA (see above for HLA typing results).
Q3: Identify the top peptide for each patient HLA allele (see above). Hint: You can download a CSV formated result file for all predictions and use R or a spreadsheet application to identify the top ranked peptides for each allele. Which approach would you rather use if you were going to scale to 100s of analysis tasks?


## 3. Identifying tumor specific peptides

Any peptide resulting from a mutation should be checked if it is present in other proteins of the human genome, which would mean that it is not specific to the tumor.

Q4: Are any of your top ranked peptides unique to the tumor? Hint: Use NCBI-BLAST to check your top identified peptides against Human sequences in the NR database.
Q5: Which peptide would you choose for potential vaccine design and why?

```{r}
library(bio3d)

## Read KRas oncogene sequence from UniProt
wt <- get.seq("P01116")

## Here we make four mutants namely: G12V, Q22N, T74S and A130V
mutant <- wt
mutant$ali[ c(12,22,74,130)] <- c("V", "N", "S", "V")

write.fasta( seqbind(wt, mutant), ids=c("wt","mutant"), file="kras-sequences.fa")
```


Q4: Are any of your top ranked peptides unique to the tumor? Hint: Use NCBI-BLAST to check your top identified peptides against Human sequences in the NR database.
Q5: Which peptide would you choose for potential vaccine design and why?



END

